File: app/src/main/java/com/surya/shopcart/FlyerAdapter.java
package com.surya.shopcart;

import static com.google.firebase.storage.FirebaseStorage.*;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

public class FlyerAdapter extends RecyclerView.Adapter<FlyerAdapter.ViewPagerHolder> {

    FirebaseStorage storage = FirebaseStorage.getInstance();
    private ArrayList<Flyer> flyersList;
    public FlyerAdapter(ArrayList<Flyer> flyersList){
        this.flyersList = flyersList;
    }
    @NonNull
    @Override
    public ViewPagerHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.flyer_layout, parent, false);
        return new FlyerAdapter.ViewPagerHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewPagerHolder holder, int position) {
        Flyer flyer = flyersList.get(position);

        StorageReference storageRef = storage.getReference();
        StorageReference pathReference = storageRef.child(flyer.getImage());

        pathReference.getDownloadUrl().addOnSuccessListener(uri -> {
            Picasso.get().load(uri).into(holder.image);
        }).addOnFailureListener(exception -> {
            Log.e("TAG", "Failed to load image from Firebase Storage", exception);
        });
    }

    @Override
    public int getItemCount() {
        return flyersList.size();
    }

    class ViewPagerHolder extends RecyclerView.ViewHolder{
        ImageView image;

        public ViewPagerHolder(View view){
            super(view);
            image = itemView.findViewById(R.id.flyerImage);
        }
    }
}

File: app/src/main/java/com/surya/shopngo/FlyerAdapter.kt
package com.surya.shopngo

import android.net.Uri
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.storage.FirebaseStorage
import com.squareup.picasso.Picasso
import com.surya.shopngo.FlyerAdapter.ViewPagerHolder

class FlyerAdapter(private val flyersList: ArrayList<Flyer>) :
    RecyclerView.Adapter<ViewPagerHolder>() {
    var storage = FirebaseStorage.getInstance()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewPagerHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.flyer_layout, parent, false)
        return ViewPagerHolder(view)
    }

    override fun onBindViewHolder(holder: ViewPagerHolder, position: Int) {
        val flyer: Flyer = flyersList[position]
        val storageRef = storage.getReference()
        val pathReference = storageRef.child(flyer.fetchImage())
        pathReference.getDownloadUrl()
            .addOnSuccessListener { uri: Uri? -> Picasso.get().load(uri).into(holder.image) }
            .addOnFailureListener { exception: Exception? ->
                Log.e(
                    "TAG",
                    "Failed to load image from Firebase Storage",
                    exception
                )
            }
    }

    override fun getItemCount(): Int {
        return flyersList.size
    }

    inner class ViewPagerHolder(view: View?) : RecyclerView.ViewHolder(view!!) {
        var image: ImageView

        init {
            image = itemView.findViewById(R.id.flyerImage)
        }
    }
}
